[identify] - No static method encodeHexString

Hello,

I have an issue while trying to use `identify` method of the `TikTokBusinessSdk`.
I'm currently using version 1.2.6 of this library, and my `targetSdkVersion` and `compileSdkVersion` are 30.
I've set up the project following the documentation [here](https://ads.tiktok.com/marketing_api/docs?id=1701890981466113)

I got this error in console :
```
     Caused by: java.lang.NoSuchMethodError: No static method encodeHexString([B)Ljava/lang/String; in class Lorg/apache/commons/codec/binary/Hex; or its super classes (declaration of 'org.apache.commons.codec.binary.Hex' appears in /system/framework/org.apache.http.legacy.jar)
        at org.apache.commons.codec.digest.DigestUtils.sha256Hex(DigestUtils.java:722)
        at com.tiktok.appevents.TTUserInfo.toSha256(TTUserInfo.java:41)
        at com.tiktok.appevents.TTUserInfo.setExternalId(TTUserInfo.java:55)
        at com.tiktok.appevents.TTAppEventLogger.identify(TTAppEventLogger.java:181)
        at com.tiktok.TikTokBusinessSdk.identify(TikTokBusinessSdk.java:466)
```

I tried to add `implementation 'commons-codec:commons-codec:1.15'` in my build.gradle, but it doesn't fix the problem.

It seems that there are some compatibility issues between Apache DigestUtils and Android (see [here](https://stackoverflow.com/a/9284092) - the post is old, but the problem still seems to be here).

Thank you !

--- a/business-core/src/main/java/com/tiktok/appevents/TTUserInfo.java
+++ b/business-core/src/main/java/com/tiktok/appevents/TTUserInfo.java
@@ -10,10 +10,10 @@ import android.content.Context;
 
 import com.tiktok.util.TTUtil;
 
-import org.apache.commons.codec.digest.DigestUtils;
 import org.json.JSONObject;
 
 import java.io.Serializable;
+import java.security.MessageDigest;
 
 public class TTUserInfo implements Cloneable, Serializable {
     static final String TAG = TTUserInfo.class.getName();
@@ -37,8 +37,19 @@ public class TTUserInfo implements Cloneable, Serializable {
     }
 
     private String toSha256(String str) {
-        if (str != null) {
-            return DigestUtils.sha256Hex(str);
+        if (str == null) {
+            return null;
+        }
+        try {
+            MessageDigest md = MessageDigest.getInstance("SHA-256");
+            md.update(str.getBytes());
+            StringBuilder result = new StringBuilder();
+            for (byte b : md.digest()) {
+                result.append(Integer.toString((b & 0xff) + 0x100, 16).substring(1));
+            }
+            return result.toString();
+        } catch (Exception e) {
+            TTCrashHandler.handleCrash(TAG, e);
         }
         return null;
     }

--- a/gradle.properties
+++ b/gradle.properties
@@ -11,7 +11,7 @@ android.enableJetifier=true
 
 GROUP=com.tiktok
 
-VERSION_NAME=1.2.6
+VERSION_NAME=1.2.7
 
 mavenGroupId = com.tiktok
 mavenArtifactId = tiktok-business-android-sdk

--- a/samples/TestApp/build.gradle
+++ b/samples/TestApp/build.gradle
@@ -7,13 +7,13 @@
 apply plugin: 'com.android.application'
 
 android {
-    compileSdkVersion 30
+    compileSdkVersion 31
     buildToolsVersion "30.0.2"
 
     defaultConfig {
         applicationId "com.example"
         minSdkVersion 17
-        targetSdkVersion 30
+        targetSdkVersion 31
         versionCode 1
         versionName "0.0.1-alpha"
 
@@ -42,18 +42,20 @@ android {
 dependencies {
     implementation fileTree(dir: "libs", include: ["*.jar"])
     implementation project(':business-core')
-    implementation 'androidx.appcompat:appcompat:1.3.1'
-    implementation 'androidx.constraintlayout:constraintlayout:2.1.1'
+    implementation 'androidx.appcompat:appcompat:1.4.0'
+    implementation 'androidx.constraintlayout:constraintlayout:2.1.2'
     implementation 'androidx.legacy:legacy-support-v4:1.0.0'
     implementation 'com.google.android.material:material:1.4.0'
     implementation 'androidx.navigation:navigation-fragment:2.3.5'
     implementation 'androidx.navigation:navigation-ui:2.3.5'
-    implementation 'androidx.lifecycle:lifecycle-process:2.3.1'
-    implementation 'androidx.lifecycle:lifecycle-common-java8:2.3.1'
+    implementation 'androidx.lifecycle:lifecycle-process:2.4.0'
+    implementation 'androidx.lifecycle:lifecycle-common-java8:2.4.0'
 
     implementation 'com.android.billingclient:billing:3.0.3'
 
     implementation 'androidx.room:room-runtime:2.3.0'
+    implementation 'androidx.lifecycle:lifecycle-livedata-ktx:2.4.0'
+    implementation 'androidx.lifecycle:lifecycle-viewmodel-ktx:2.4.0'
     annotationProcessor 'androidx.room:room-compiler:2.3.0'
 
     testImplementation 'junit:junit:4.13.2'

--- a/samples/TestApp/src/main/AndroidManifest.xml
+++ b/samples/TestApp/src/main/AndroidManifest.xml
@@ -6,18 +6,19 @@
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~-->
 
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.example">
+          xmlns:tools="http://schemas.android.com/tools" package="com.example">
 
     <uses-permission android:name="android.permission.INTERNET" />
     <uses-permission android:name="com.android.vending.BILLING" />
 
     <application
-        android:allowBackup="true"
-        android:icon="@mipmap/ic_launcher"
-        android:label="@string/app_name"
-        android:roundIcon="@mipmap/ic_launcher_round"
-        android:supportsRtl="true"
-        android:theme="@style/AppTheme.NoActionBar">
+            android:allowBackup="true"
+            android:icon="@mipmap/ic_launcher"
+            android:label="@string/app_name"
+            android:roundIcon="@mipmap/ic_launcher_round"
+            android:supportsRtl="true"
+            android:networkSecurityConfig="@xml/network_security_config"
+            android:theme="@style/AppTheme.NoActionBar" tools:targetApi="n">
         <activity
             android:name=".ui.events.PropEditActivity"
             android:theme="@style/AppTheme" />

--- a/samples/TestApp/src/main/java/com/example/App.java
+++ b/samples/TestApp/src/main/java/com/example/App.java
@@ -12,9 +12,11 @@ import android.os.Looper;
 import android.view.Menu;
 import android.view.MenuItem;
 
+import androidx.lifecycle.ViewModelProvider;
 import com.example.model.EventLog;
 import com.example.persistence.EventLogRepo;
 import com.example.testdata.TestEvents;
+import com.example.ui.home.HomeViewModel;
 import com.google.android.material.navigation.NavigationView;
 
 import androidx.annotation.NonNull;
@@ -38,6 +40,7 @@ import java.util.Objects;
 public class App extends AppCompatActivity {
 
     private AppBarConfiguration mAppBarConfiguration;
+    private HomeViewModel homeViewModel;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
@@ -55,6 +58,7 @@ public class App extends AppCompatActivity {
         NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment);
         NavigationUI.setupActionBarWithNavController(this, navController, mAppBarConfiguration);
         NavigationUI.setupWithNavController(navigationView, navController);
+        homeViewModel = new ViewModelProvider(this).get(HomeViewModel.class);
 
         if (savedInstanceState == null) {
             // !!!!!!!!!!!!!!!!!!!!!!!!!
@@ -75,7 +79,10 @@ public class App extends AppCompatActivity {
             long afterStartTT = System.currentTimeMillis();
             android.util.Log.i("TikTokBusinessSdk", " Init time in ms " + (afterStartTT-beforeStartTT));
 
-            // testing delay tracking, implementing a 60 sec delay manually
+            // check if user info is cached & init
+            homeViewModel.checkInitTTAM();
+
+            // testing delay tracking, implementing a 6 sec delay manually
             // ideally has to be after accepting tracking permission
             new Handler(Looper.getMainLooper()).postDelayed(TikTokBusinessSdk::startTrack, 6000);
         }

--- a/samples/TestApp/src/main/java/com/example/ui/home/HomeFragment.java
+++ b/samples/TestApp/src/main/java/com/example/ui/home/HomeFragment.java
@@ -43,9 +43,7 @@ public class HomeFragment extends Fragment {
         homeViewModel = new ViewModelProvider(this).get(HomeViewModel.class);
         View root = inflater.inflate(R.layout.fragment_home, container, false);
         final TextView textView = root.findViewById(R.id.text_home);
-        homeViewModel.getText().observe(getViewLifecycleOwner(), s -> {
-            textView.setText(s);
-        });
+        homeViewModel.getText().observe(getViewLifecycleOwner(), s -> textView.setText(s));
 
         textView.setOnClickListener(v -> {
             if (purchase != null) {

--- a/samples/TestApp/src/main/java/com/example/ui/home/HomeViewModel.java
+++ b/samples/TestApp/src/main/java/com/example/ui/home/HomeViewModel.java
@@ -6,17 +6,33 @@
 
 package com.example.ui.home;
 
+import android.app.Application;
+import android.content.Context;
+import android.content.SharedPreferences;
+import androidx.annotation.Nullable;
+import androidx.lifecycle.AndroidViewModel;
 import androidx.lifecycle.LiveData;
 import androidx.lifecycle.MutableLiveData;
-import androidx.lifecycle.ViewModel;
+import com.example.R;
+import com.tiktok.TikTokBusinessSdk;
 
-public class HomeViewModel extends ViewModel {
+public class HomeViewModel extends AndroidViewModel {
 
-    private MutableLiveData<String> mText;
+    private final MutableLiveData<String> mText;
+    SharedPreferences sharedPreferences;
 
-    public HomeViewModel() {
+    public HomeViewModel(Application application) {
+        super(application);
         mText = new MutableLiveData<>();
         mText.setValue("Purchase");
+        sharedPreferences = getApplication().getSharedPreferences("TT_IDENTIFY", Context.MODE_PRIVATE);
+    }
+
+    public void checkInitTTAM() {
+        if (checkAMCacheExist()) {
+            TikTokBusinessSdk.identify(getExternalID(), getExternalUsername(),
+                    getExternalPhoneNumber(), getExternalEmail());
+        }
     }
 
     public void setText(String txt) {
@@ -25,5 +41,54 @@ public class HomeViewModel extends ViewModel {
 
     public LiveData<String> getText() {
         return mText;
+    }
+
+    public void setNewCache(String externalId,
+                            @Nullable String externalUserName,
+                            @Nullable String phoneNumber,
+                            @Nullable String email) {
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        editor.putString(getApplication().getString(R.string.tt_am_external_id), externalId);
+        if (externalUserName != null) {
+            editor.putString(getApplication().getString(R.string.tt_am_external_username), externalUserName);
+        }
+        if (phoneNumber != null) {
+            editor.putString(getApplication().getString(R.string.tt_am_external_phone), phoneNumber);
+        }
+        if (email != null) {
+            editor.putString(getApplication().getString(R.string.tt_am_external_email), email);
+        }
+        editor.putBoolean(getApplication().getString(R.string.tt_am_init), true);
+        editor.apply();
+    }
+
+    public void resetCache() {
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        editor.remove(getApplication().getString(R.string.tt_am_external_id));
+        editor.remove(getApplication().getString(R.string.tt_am_external_username));
+        editor.remove(getApplication().getString(R.string.tt_am_external_phone));
+        editor.remove(getApplication().getString(R.string.tt_am_external_email));
+        editor.remove(getApplication().getString(R.string.tt_am_init));
+        editor.apply();
+    }
+
+    boolean checkAMCacheExist() {
+        return sharedPreferences.getBoolean(getApplication().getString(R.string.tt_am_init), false);
+    }
+
+    String getExternalID() {
+        return sharedPreferences.getString(getApplication().getString(R.string.tt_am_external_id), null);
+    }
+
+    String getExternalUsername() {
+        return sharedPreferences.getString(getApplication().getString(R.string.tt_am_external_username), null);
+    }
+
+    String getExternalPhoneNumber() {
+        return sharedPreferences.getString(getApplication().getString(R.string.tt_am_external_phone), null);
+    }
+
+    String getExternalEmail() {
+        return sharedPreferences.getString(getApplication().getString(R.string.tt_am_external_email), null);
     }
 }

--- a/samples/TestApp/src/main/java/com/example/ui/identify/IdentifyFragment.java
+++ b/samples/TestApp/src/main/java/com/example/ui/identify/IdentifyFragment.java
@@ -1,0 +1,68 @@
+/*******************************************************************************
+ * Copyright (c) 2021. Bytedance Inc.
+ *
+ * This source code is licensed under the MIT license found in the LICENSE file in the root directory of this source tree.
+ ******************************************************************************/
+
+package com.example.ui.identify;
+
+import android.widget.EditText;
+import android.widget.TextView;
+import androidx.lifecycle.ViewModelProvider;
+import android.os.Bundle;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.fragment.app.Fragment;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import com.example.R;
+import com.example.ui.home.HomeViewModel;
+import com.tiktok.TikTokBusinessSdk;
+
+public class IdentifyFragment extends Fragment {
+
+    private HomeViewModel mViewModel;
+
+    public static IdentifyFragment newInstance() {
+        return new IdentifyFragment();
+    }
+
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
+                             @Nullable Bundle savedInstanceState) {
+
+        mViewModel = new ViewModelProvider(this).get(HomeViewModel.class);
+        View root = inflater.inflate(R.layout.identify_fragment, container, false);
+
+        root.findViewById(R.id.resetBtn).setOnClickListener(view -> {
+            TikTokBusinessSdk.logout();
+            mViewModel.resetCache();
+        });
+
+        root.findViewById(R.id.identifyBtn).setOnClickListener(view -> {
+            EditText externalID = root.findViewById(R.id.externalID);
+            EditText externalUsername = root.findViewById(R.id.externalUsername);
+            EditText externalPhone = root.findViewById(R.id.externalPhone);
+            EditText externalEmail = root.findViewById(R.id.externalEmail);
+            String externalIDStr = externalID.getText().toString();
+            String externalUsernameStr = externalUsername.getText().toString();
+            String externalPhoneStr = externalPhone.getText().toString();
+            String externalEmailStr = externalEmail.getText().toString();
+            if (externalUsernameStr.equals("")) externalUsernameStr = null;
+            if (externalPhoneStr.equals("")) externalPhoneStr = null;
+            if (externalEmailStr.equals("")) externalEmailStr = null;
+            TikTokBusinessSdk.identify(externalIDStr, externalUsernameStr, externalPhoneStr, externalEmailStr);
+            mViewModel.setNewCache(externalIDStr, externalUsernameStr, externalPhoneStr, externalEmailStr);
+        });
+
+        root.findViewById(R.id.logoutBtn).setOnClickListener(view -> TikTokBusinessSdk.logout());
+        return root;
+    }
+
+    @Override
+    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+        super.onActivityCreated(savedInstanceState);
+    }
+
+}

--- a/samples/TestApp/src/main/res/layout/identify_fragment.xml
+++ b/samples/TestApp/src/main/res/layout/identify_fragment.xml
@@ -1,0 +1,168 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ~ Copyright (c) 2021. Bytedance Inc.
+  ~
+  ~ This source code is licensed under the MIT license found in the LICENSE file in the root directory of this source tree.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~-->
+
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+                                                   xmlns:tools="http://schemas.android.com/tools"
+                                                   android:layout_width="match_parent"
+                                                   android:layout_height="match_parent"
+                                                   xmlns:app="http://schemas.android.com/apk/res-auto"
+                                                   tools:context=".ui.identify.IdentifyFragment">
+
+    <LinearLayout
+            android:id="@+id/externalIDLayout"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:gravity="center"
+            android:orientation="horizontal"
+            android:paddingTop="10dp"
+            tools:ignore="MissingConstraints">
+
+        <TextView
+                android:id="@+id/externalIDLabel"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:labelFor="@+id/externalID"
+                android:text="@string/external_id"
+                android:textColor="@color/colorLabel" />
+
+        <EditText
+                android:id="@+id/externalID"
+                android:layout_width="180dp"
+                android:layout_height="wrap_content"
+                android:layout_marginStart="20dp"
+                android:autofillHints="@string/tt_am_external_id"
+                android:inputType="text"
+                android:singleLine="true" />
+
+    </LinearLayout>
+
+    <LinearLayout
+            android:id="@+id/externalUsernameLayout"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:gravity="center"
+            android:orientation="horizontal"
+            android:paddingTop="10dp"
+            app:layout_constraintTop_toBottomOf="@+id/externalIDLayout"
+            tools:ignore="MissingConstraints">
+
+        <TextView
+                android:id="@+id/externalUsernameLabel"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:labelFor="@+id/externalUsername"
+                android:text="@string/username"
+                android:textColor="@color/colorLabel" />
+
+        <EditText
+                android:id="@+id/externalUsername"
+                android:layout_width="180dp"
+                android:layout_height="wrap_content"
+                android:layout_marginStart="20dp"
+                android:autofillHints="@string/username"
+                android:inputType="text"
+                android:singleLine="true" />
+
+    </LinearLayout>
+
+    <LinearLayout
+            android:id="@+id/externalPhoneLayout"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:gravity="center"
+            android:orientation="horizontal"
+            android:paddingTop="10dp"
+            app:layout_constraintTop_toBottomOf="@+id/externalUsernameLayout"
+            tools:ignore="MissingConstraints">
+
+        <TextView
+                android:id="@+id/externalPhoneLabel"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:labelFor="@+id/externalPhone"
+                android:text="@string/phone"
+                android:textColor="@color/colorLabel" />
+
+        <EditText
+                android:id="@+id/externalPhone"
+                android:layout_width="180dp"
+                android:layout_height="wrap_content"
+                android:layout_marginStart="20dp"
+                android:autofillHints="@string/phone"
+                android:inputType="text"
+                android:singleLine="true" tools:ignore="TextFields"/>
+
+    </LinearLayout>
+
+    <LinearLayout
+            android:id="@+id/externalEmailLayout"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:gravity="center"
+            android:orientation="horizontal"
+            android:paddingTop="10dp"
+            app:layout_constraintTop_toBottomOf="@+id/externalPhoneLayout"
+            tools:ignore="MissingConstraints">
+
+        <TextView
+                android:id="@+id/externalEmailLabel"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:labelFor="@+id/externalEmail"
+                android:text="@string/email"
+                android:textColor="@color/colorLabel" />
+
+        <EditText
+                android:id="@+id/externalEmail"
+                android:layout_width="180dp"
+                android:layout_height="wrap_content"
+                android:layout_marginStart="20dp"
+                android:autofillHints="@string/email"
+                android:inputType="text"
+                android:singleLine="true" tools:ignore="TextFields"/>
+
+    </LinearLayout>
+
+    <RelativeLayout
+            android:id="@+id/footer"
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:gravity="bottom"
+            tools:ignore="MissingConstraints">
+
+        <LinearLayout
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_alignParentBottom="true"
+                android:layout_marginBottom="10dp"
+                android:gravity="center"
+                android:orientation="horizontal" tools:ignore="MissingConstraints">
+
+            <Button
+                    android:id="@+id/resetBtn"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:padding="5dp"
+                    android:text="@string/reset" />
+
+            <Button
+                    android:id="@+id/identifyBtn"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:padding="5dp"
+                    android:text="@string/identify" />
+
+            <Button
+                    android:id="@+id/logoutBtn"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:padding="5dp"
+                    android:text="@string/logout" />
+        </LinearLayout>
+    </RelativeLayout>
+
+</androidx.constraintlayout.widget.ConstraintLayout>

--- a/samples/TestApp/src/main/res/menu/activity_main_drawer.xml
+++ b/samples/TestApp/src/main/res/menu/activity_main_drawer.xml
@@ -15,12 +15,16 @@
             android:icon="@drawable/ic_baseline_view_list_24"
             android:title="@string/menu_events" />
         <item
+                android:id="@+id/nav_identify"
+                android:icon="@drawable/ic_baseline_view_list_24"
+                android:title="@string/menu_identify" />
+        <item
             android:id="@+id/nav_eventlog"
             android:icon="@drawable/ic_baseline_view_list_24"
             android:title="@string/menu_eventlog" />
         <item
             android:id="@+id/nav_billing"
-            android:icon="@drawable/ic_baseline_home_24"
+            android:icon="@drawable/ic_baseline_view_list_24"
             android:title="@string/menu_billing" />
     </group>
 </menu>

--- a/samples/TestApp/src/main/res/navigation/mobile_navigation.xml
+++ b/samples/TestApp/src/main/res/navigation/mobile_navigation.xml
@@ -28,4 +28,10 @@
         android:name="com.example.ui.eventlog.EventLogFragment"
         android:label="@string/menu_eventlog"
         tools:layout="@layout/fragment_eventlog" />
+
+    <fragment
+            android:id="@+id/nav_identify"
+            android:name="com.example.ui.identify.IdentifyFragment"
+            android:label="@string/menu_identify"
+            tools:layout="@layout/identify_fragment" />
 </navigation>

--- a/samples/TestApp/src/main/res/values/strings.xml
+++ b/samples/TestApp/src/main/res/values/strings.xml
@@ -22,9 +22,22 @@
 
     <string name="menu_billing">Google play Billing V4</string>
     <string name="menu_events">Events</string>
+    <string name="menu_identify">Identify</string>
     <string name="menu_eventlog">Event Logs</string>
     <string name="properties">Event Properties</string>
     <string name="clear_log">Clear Log</string>
     <string name="load_test">Load Test</string>
     <string name="preload_all_events">Preload All Events</string>
+
+    <string name="tt_am_init">TT_AM_INIT</string>
+    <string name="tt_am_external_id">TT_AM_EXTERNAL_ID</string>
+    <string name="tt_am_external_username">TT_AM_EXTERNAL_USERNAME</string>
+    <string name="tt_am_external_phone">TT_AM_EXTERNAL_PHONE</string>
+    <string name="tt_am_external_email">TT_AM_EXTERNAL_EMAIL</string>
+    <string name="external_id">External ID</string>
+    <string name="username">Username</string>
+    <string name="phone">Phone</string>
+    <string name="email">Email</string>
+    <string name="identify">Identify</string>
+    <string name="logout">Logout</string>
 </resources>

--- a/samples/TestApp/src/main/res/xml/network_security_config.xml
+++ b/samples/TestApp/src/main/res/xml/network_security_config.xml
@@ -1,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ~ Copyright (c) 2021. Bytedance Inc.
+  ~
+  ~ This source code is licensed under the MIT license found in the LICENSE file in the root directory of this source tree.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~-->
+
+<network-security-config>
+    <base-config cleartextTrafficPermitted="true">
+        <trust-anchors>
+            <certificates src="system"/>
+        </trust-anchors>
+    </base-config>
+    <debug-overrides>
+        <trust-anchors>
+            <certificates src="user"/>
+        </trust-anchors>
+    </debug-overrides>
+</network-security-config>

--- a/samples/internalmonitor/build.gradle
+++ b/samples/internalmonitor/build.gradle
@@ -17,7 +17,7 @@ android {
 
     defaultConfig {
         applicationId "com.example.internalmonitor"
-        minSdkVersion 16
+        minSdkVersion 17
         targetSdkVersion 30
         versionCode 3
         versionName "3.0"

